Makefile

```
@@ -0,0 +1,57 @@
+# Copyright (c) 2022 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+REGISTRY          := ghcr.io/oliver-goetz
+PUSH_LATEST_TAG   := true
+VERSION           := $(shell cat VERSION)
+EFFECTIVE_VERSION := $(VERSION)-$(shell git rev-parse --short HEAD)
+OS                := linux
+ARCH              := amd64
+
+IMG_CLA_ASSISTANT := cla-assistant
+REG_CLA_ASSISTANT := $(REGISTRY)/$(IMG_CLA_ASSISTANT)
+
+
+#################################################################
+# Rules related to binary build, Docker image build and release #
+#################################################################
+.PHONY: docker-images
+docker-images:
+ifeq ("$(REGISTRY)", "")
+	@echo "Please set your docker registry in .REGISTRY file first."; false;
+endif
+	@echo "Building docker images with version and tag $(EFFECTIVE_VERSION)"
+	@docker build --build-arg EFFECTIVE_VERSION=$(EFFECTIVE_VERSION) --build-arg ARCH=$(ARCH) --build-arg OS=$(OS) -t $(REG_CLA_ASSISTANT):$(EFFECTIVE_VERSION) -t $(REG_CLA_ASSISTANT):latest -f Dockerfile --target $(IMG_CLA_ASSISTANT) .
+
+.PHONY: docker-push
+docker-push:
+ifeq ("$(REGISTRY)", "")
+	@echo "Please set your docker registry in .REGISTRY file first."; false;
+endif
+	@if ! docker images $(REG_CLA_ASSISTANT) | awk '{ print $$2 }' | grep -q -F $(EFFECTIVE_VERSION); then echo "$(REG_CLA_ASSISTANT) version $(EFFECTIVE_VERSION) is not yet built. Please run 'make docker-images'"; false; fi
+	@docker push $(REG_CLA_ASSISTANT):$(EFFECTIVE_VERSION)
+ifeq ("$(PUSH_LATEST_TAG)", "true")
+	@docker push $(REG_CLA_ASSISTANT):latest
+endif
+
+
+#####################################################################
+# Rules for verification, formatting, linting, testing and cleaning #
+#####################################################################
+
+.PHONY: revendor
+revendor:
+	@GO111MODULE=on go mod tidy
+	@GO111MODULE=on go mod vendor
```

```suggestion
modules:
	go mod tidy
```

Let's not vendor in this repository, as long as there is no good reason for it. Vendoring can be quite painful, think git operation performance or GitHub reviews (this review is super hard because GitHub UI is lagging so much...)

---

The kubernetes/test-infra is changing quite fast, so my gut feeling is that not every dependency will be resolvable at any time. Additionally, it builds much faster. That's why I started it. 

A review with 4666 is painful indeed. I could remove it for the review and reintroduce it later. There is not so much for reviewing on vendoring anyhow. What do you think?

---

> my gut feeling is that not every dependency will be resolvable at any time

Hmm, if I understood go modules, goproxy, etc. correctly (which might not be the case, because it's highly confusing) all module references should be always be resolvable later on.

> Additionally, it builds much faster

Hmm, if done properly (e.g. caching the go build cache in containers, etc.) not vendoring should be even faster, I think. Also, you will profit from a shared build cache between different projects, preventing a lot of file copying back and forth (sending build context to docker daemon or other build tool, git operations, etc.).

> A review with 4666 changed files is painful indeed. I could remove it for the review and reintroduce it later. There is not so much for reviewing on vendoring anyhow. What do you think?

If we stick to vendoring at the end, you could simply move the `vendor/` changes to dedicated and isolated commits. That's what I usually do for larger PRs in g/g. It very much helps the reviewers :)